pkg/controller/iam/serviceaccountkey.go

```diff
@@ -0,0 +1,274 @@
+		KeyAlgorithm:   gcp.StringValue(cr.Spec.ForProvider.KeyAlgorithm),
+		PrivateKeyType: gcp.StringValue(cr.Spec.ForProvider.PrivateKeyType),
+	})
+
+	fromProvider, err := req.Context(ctx).Do()
+	if err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreateServiceAccountKey)
+	}
+	connDetails, err := getConnectionDetails(cr.Spec.ForProvider.PublicKeyType, fromProvider)
+	if err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreateServiceAccountKey)
+	}
+	keyID, err := serviceaccountkey.ParseKeyIDFromRrn(fromProvider.Name)
+	if err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreateServiceAccountKey)
+	}
+
+	meta.SetExternalName(cr, keyID) // set external name to key id parsing it from Google Cloud API relative resource name
+
+	return managed.ExternalCreation{
+		ExternalNameAssigned: true,
+		ConnectionDetails:    connDetails,
+	}, nil
+}
+
+func (s *serviceAccountKeyExternalClient) Update(_ context.Context,
+	_ resource.Managed) (managed.ExternalUpdate, error) {
+	// ServiceAccountKeys are immutable, i.e.,GCP IAM Rest API does not provide an update method:
+	// https://cloud.google.com/iam/docs/reference/rest/v1/projects.serviceAccounts.keys
+	return managed.ExternalUpdate{}, nil
+}
+
+func (s *serviceAccountKeyExternalClient) Delete(ctx context.Context, mg resource.Managed) error {
+	cr, ok := mg.(*v1alpha1.ServiceAccountKey)
+
+	if !ok {
+		return errors.New(errNotServiceAccountKey)
+	}
+
+	rrn, _, err := resourcePath(cr)
+
+	if err != nil {
+		return errors.Wrap(err, errDeleteServiceAccountKey)
+	}
+
+	_, err = s.serviceAccountKeyClient.Delete(rrn).Context(ctx).Do()
+
+	return errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errDeleteServiceAccountKey)
+}
+
+// referencedServiceAccountPath returns the external name of the service account which saKey belongs to
+func referencedServiceAccountPath(saKey *v1alpha1.ServiceAccountKey) (string, error) {
+	// we assume Google Cloud API relative resource name (aka resource path) of the ServiceAccount is stored in spec.ForProvider.ServiceAccount
+	if saKey.Spec.ForProvider.ServiceAccount == nil || *saKey.Spec.ForProvider.ServiceAccount == "" {
+		return "", fmt.Errorf(fmtErrInvalidServiceAccountRef, saKey.Spec.ForProvider.ServiceAccountReferer)
+	}
+
+	return *saKey.Spec.ForProvider.ServiceAccount, nil
+}
+
+// resourcePath yields the Google Cloud API relative resource name for the ServiceAccountKey resource
+//   returns <the relative resource name>, <whether the external name annotation is non-empty>, <error encountered during resolution>
+func resourcePath(saKey *v1alpha1.ServiceAccountKey) (string, bool, error) {
+	if saPath, err := referencedServiceAccountPath(saKey); err != nil {
+		return "", false, err
+	} else if extName := meta.GetExternalName(saKey); extName == "" {
+		return "", false, nil
+	} else {
+		return fmt.Sprintf(fmtKeyRelativeResourceName, saPath, extName), true, nil
+	}
+}
```

Comment:
```suggestion
func resourcePath(saKey *v1alpha1.ServiceAccountKey) string {
	sa := gcp.StringValue(saKey.Spec.ForProvider.ServiceAccount)
	if sa == "" {
		return ""
	}
	return fmt.Sprintf(fmtKeyRelativeResourceName, sa, meta.GetExternalName(saKey))
}
```

Would this simplification be worth it? It seems we have access to `meta.GetExternalName(saKey)` for emptiness check in `Observe` and `gcp.StringValue(saKey.Spec.ForProvider.ServiceAccount)` would work in places that require `referencedServiceAccountPath(saKey)`, i.e. we'd return GCP error to users saying it's empty rather than us. Even only `return fmt.Sprintf(fmtKeyRelativeResourceName, gcp.StringValue(saKey.Spec.ForProvider.ServiceAccount), meta.GetExternalName(saKey))` could be enough but I'm not sure whether an unambiguous error would be returned in that case, i.e. `sa name is not given` instead of `request malformed`.